File: practice/leetcode/cpp/102_binary_tree_level_order_traversal.cpp
//https://leetcode.com/problems/binary-tree-level-order-traversal/description/
//Given the root of a binary tree, return the level order traversal of its nodes' values. 
//(i.e., from left to right, level by level).

#include <queue>
#include <utility>
#include <vector>
using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(): val(0), left(nullptr), right(nullptr) {}
	TreeNode(int x): val(x), left(nullptr), right(nullptr) {}
	TreeNode(int x, TreeNode *left, TreeNode *right): val(x), left(left), right(right) {}
};

class Solution {
public:
	vector<vector<int>> levelOrder(TreeNode *root) {
		if (!root) {
			return {};
		} 
		
		vector<vector<int>> answer;
		queue<TreeNode *> que;
		TreeNode *temp;

		que.push(root);
		while (!que.empty()) {
			vector<int> level;

			for (int i = que.size(); i; i--) {
				level.push_back(que.front()->val);
				temp = que.front();
				que.pop();
				if (temp->left) {
					que.push(temp->left);
				}
				if (temp->right) {
					que.push(temp->right);
				}
			}
			answer.push_back(std::move(level));
 		}
  		return(answer);
	}
};

File: practice/leetcode/python/102_binary_tree_level_order_traversal.py
#https://leetcode.com/problems/binary-tree-level-order-traversal/description/
#Given the root of a binary tree, return the level order traversal of its nodes' values. 
#(i.e., from left to right, level by level).
from typing import Optional
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        answer: List[List[int]] = []
        queue = []
        queue.append(root)

        while len(queue) > 0:
            level = []
            for _ in range(len(queue)):
                node = queue.pop(0)
                level.append(node.val)
                if (node.left):
                    queue.append(node.left)
                if (node.right):
                    queue.append(node.right)

            answer.append(level)
        return(answer)

    def levelOrder2(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        answer = List[List[int]] = []
        queue = []
        queue.append(root)

        while len(queue):
            level = []
            for _ in range(len(queue)):
                node = queue.pop(0)
                level.appned(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            answer.append(level)
        return(answer)

    def levelOrder3(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        answer = []
        queue = []
        queue.append(root)

        while len(queue):
            level = []
            for _ in range(len(queue)):
                node = queue.pop(0)
                level.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            answer.append(level)
        return(answer)
